lib/vrl/compiler/src/value/target.rs

```diff
@@ -342,60 +341,68 @@ impl Value {
                 handle_field(field, new)
             }
 
-            Index(index) => match self {
-                // If the current value is an array, we need to either swap out
-                // an existing value, or append the new value to the array.
-                Value::Array(array) => {
-                    let index = *index;
+            Index(index) => {
+                let array = match self {
+                    Value::Array(array) => array,
+                    _ => {
+                        *self = Value::Array(vec![]);
+                        self.as_array_mut().unwrap()
+                    }
+                };
+
+                let index = *index;
+
+                // If we're dealing with a negative index, we either need to
+                // replace an existing value, or insert to the front of the
+                // array.
+                if index.is_negative() {
+                    let abs = index.abs() as usize;
 
-                    // If the array has less items than needed, we'll fill it in
-                    // with `Null` values.
-                    if index > 0i64 && array.len() < index as usize {
-                        array.resize(index as usize, Value::Null);
+                    // left-padded with null values
+                    for _ in 1..abs - array.len() {
+                        array.insert(0, Value::Null)
```

Comment:
I guess since we are only expecting to deal with very small arrays, this is ok. But I wonder if it would be faster to create a Vector with the correct number of elements in it and prepend that to this one..

Reply:
I agree. I didn't want to worry about performance here just yet, but obviously anything that has to push to the front of the array is an expensive operation. We could optimize this by allocating to the front once, or by using `VecDeque`, but I didn't want to go that route in this PR for now.